### PR TITLE
ADD: [gles] BOB deinterlacing

### PR DESCRIPTION
**Pull request by koying**

### CODE REVIEW

**Comment by FernetMenta** [comment]
awesome!

**Comment by koying** [comment]
@FernetMenta Thanks for your tips :)
Test builds coming

**Comment by fritsch** [comment]
awesome! @smallint @wolfgar ping!

**Comment by smallint** [comment]
I keep that in mind. If someone wants to test it with current mainline, I am interested in results. As we are focusing currently on getting rid of GL for rendering, we don't use shaders anymore. One could enable switching between both methods later, though. Thanks for this.

**Comment by koying** [comment]
I've added BOB for the yub2rgb path.

The easy way is 72b68761d3c178c69db8345a12fa17fd96b48ed3 , but the fact that GL_UNPACK_ROW_LENGTH is not implemented in GLES 2 has a big impact, so I added c851660faad618ef2aee0dee1f7e36893501722b to reuse an existing shader which apparently fell in disuse..

GL_UNPACK_ROW_LENGTH is implemented in GLES 3, though, but it's not quite mainstream yet. The alternative would thus be to go with the "standard" version and switch to GLES 3 in due time.

I've also reworked the shaders to remove the "if", which is apparently is less efficient in GLSL than doing unnecessary blending.

While playing with shaders, I also implemented contrast/brightness for gles. Is it ok to add to this PR? Otherwise, it will wait for this one to go in to avoid rebase hell.

**Comment by fritsch** [comment]
Add them as separate commits to this PR I'd say.

Do you have an overview of which platforms have proper GLES3 support yet? Can those two be "#fidefed" somehow - so that we can use the best out of both worlds?

**Comment by FernetMenta** [comment]
Im am fine with added commits to this PR too. Any issues in this area will be forwarded to the desk of our new shader expert anyway :)

**Comment by koying** [comment]
Re GLES3, see, e.g. http://en.wikipedia.org/wiki/Mali_(GPU)
Mali 4XX is the vast majority of GPU around atm, and does not support it.

But yeah, I could implement GLES3 and fallback to GLES2.

**Comment by fritsch** [comment]
That would be awesome. Performance wise - how fast works your implementation? Do we get 50 fps out of it?

**Comment by koying** [comment]
Performances obviously depend on the GPU, but I get 50fps out of an AFTV Stick, which is quite low end, both on HD (accelerated) and SD (ffmpeg).

**Comment by fritsch** [comment]
That's very nice! If you need some testing samples: http://solidrun.maltegrosse.de/~fritsch/

This PR is ready when you are :-)

File: xbmc/cores/VideoRenderers/LinuxRendererGLES.cpp [hunk 1]
```diff
@@ -786,21 +803,32 @@ void CLinuxRendererGLES::LoadShaders(int field)
...
+            m_pYUVProgShader->Free();
+            delete m_pYUVProgShader;
+            m_pYUVProgShader = NULL;
+          }
+          if (m_pYUVBobShader)
+          {
+            m_pYUVBobShader->Free();
+            delete m_pYUVBobShader;
+            m_pYUVBobShader = NULL;
+          }
```
Comment (Jalle19) [inline xbmc/cores/VideoRenderers/LinuxRendererGLES.cpp]: This code is now duplicated in three places, might as well make it a method

Comment (koying) [inline xbmc/cores/VideoRenderers/LinuxRendererGLES.cpp]: The whole file deserves a big refactoring, really...
Plenty of duplication between the various render methods.

**Comment by koying** [comment]
jenkins build this please

@fritsch If you want to test deinterlacing, this is the tricky sample to use ;)
http://www.semperpax.be/owncloud/public.php?service=files&t=6cc4e6fc01412e346a9ebf173a2ceca4

**Comment by fritsch** [comment]
Hehe - thanks much. I always looked at the burosch samples, which are also
quite nice. But yours rocks more :-)

2015-01-20 14:20 GMT+01:00 Chris Browet <notifications@github.com>:

> jenkins build this please
>
> @fritsch <https://github.com/fritsch> If you want to test deinterlacing,
> this is the tricky sample to use ;)
>
> http://www.semperpax.be/owncloud/public.php?service=files&t=6cc4e6fc01412e346a9ebf173a2ceca4
>
> --
> Reply to this email directly or view it on GitHub
> <https://github.com/xbmc/xbmc/pull/6169#issuecomment-70651683>.
>



-- 
                   Key-ID:     0x1A995A9B
                   keyserver: pgp.mit.edu
==============================================================
Fingerprint: 4606 DA19 EC2E 9A0B 0157  C81B DA07 CF63 1A99 5A9B

**Comment by koying** [comment]
Unrelated win32 failure, I guess? (http://jenkins.kodi.tv/job/WIN-32/3244/console)
@Memphiz ?

Will merge in 24h if no further comments.

**Comment by Memphiz** [comment]
i think that slave has network issues ...

**Comment by koying** [comment]
jenkins build and merge

**Pull request merged**